### PR TITLE
Use aligned allocs for > PAGE_SIZE allocations, avoid unaligned memcpy/memset on large buffers

### PR DESCRIPTION
**Pull request by justinboswell**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/allocator.c [hunk 1]
```diff
@@ -25,30 +25,65 @@
 #    pragma warning(disable : 4100)
 #endif
 
+#ifndef PAGE_SIZE
+#    define PAGE_SIZE (4 * 1024)
```
Comment (bretambrose) [inline source/allocator.c:29]: Is there any way we can genuinely determine this per-platform as it's probably a kernel setting?

File: source/allocator.c [hunk 2]
```diff
@@ -25,30 +25,65 @@
...
     /* An allocator must define mem_acquire and mem_release.  All other fields are optional */
     return alloc && AWS_OBJECT_PTR_IS_READABLE(alloc) && alloc->mem_acquire && alloc->mem_release;
 }
 
 static void *s_default_malloc(struct aws_allocator *allocator, size_t size) {
     (void)allocator;
-    return malloc(size);
+    /* larger allocations should be aligned so that AVX and friends can run
+     * the aligned versions of memcpy and memset and friends on big buffers */
+    const size_t alignment = sizeof(void *) * (size > PAGE_SIZE ? 8 : 1);
```
Comment (bretambrose) [inline source/allocator.c:41]: Can we add a comment for us mere mortals explicitly giving out the contract:

64 byte alignment for > page allocations on 64 bit systems
32 byte alignment for > page allocations on 32 bit systems
8 byte alignment for <= page allocations on 64 bit systems
4 byte alignment for <= page allocations on 32 bit systems

At least that's that the code does.  I'd feel more confident in the correctness if the comment echoed that contract.

File: source/allocator.c (hunk 1 above)
Comment (justinboswell) [reply source/allocator.c:29]: On POSIX, it's defined as PAGE_SIZE. On windows, it's 4K unless you twiddle it yourself.

File: source/allocator.c (hunk 2 above)
Comment (justinboswell) [reply source/allocator.c:41]: Good call.

File: source/allocator.c (hunk 1 above)
Comment (bretambrose) [reply source/allocator.c:29]: Ok, that's helpful

Comment (justinboswell) [reply source/allocator.c:29]: Additionally, you could use madvise() on POSIX to switch to HUGEPAGE, which makes 2MB pages. This is usually best left to the system to worry about. We could also encode this as a separate value that we default to PAGE_SIZE, but can be overridden. Thoughts?

**Review by bretambrose — APPROVED** [review]

File: source/allocator.c (hunk 1 above)
Comment (bretambrose) [reply source/allocator.c:29]: Brain buffer overrun to be honest.

**Pull request merged**